src/index.ts

```diff
@@ -23,57 +28,92 @@ function loadArtifact(contract: string) {
+    if (providerOrWeb3.currentProvider) {
+      this.provider = providerOrWeb3.currentProvider;
+      this.web3 = providerOrWeb3;
+    } else {
+      this.provider = providerOrWeb3;
+    }
 
-  function fromABI(abi: object, bytecode = '') {
-    // eslint-disable-next-line @typescript-eslint/camelcase
-    return new web3Contract(abi, undefined, { data: bytecode, from: defaultSender, gas: defaultGas });
+    this.defaultSender = defaultSender;
+    this.defaultGas = defaultGas;
   }
 
-  function fromArtifact(contract: string) {
+  public fromArtifact(contract: string, address?: string): any {
     const { abi, bytecode } = loadArtifact(contract);
-    return fromABI(abi, bytecode);
+    return this.fromABI(abi, bytecode, address);
   }
 
-  return { fromABI, fromArtifact };
+  public abstract fromABI(abi: object, bytecode?: string, address?: string): any;
 }
 
-function truffleLoader(provider: any, defaultSender: string, defaultGas: number) {
-  const truffleContract = tryRequire('@truffle/contract');
-  if (truffleContract === undefined) {
-    throw new Error(
-      "Could not load package '@truffle/contract'. Please install it alongisde @openzeppelin/contract-loader.",
-    );
+export class Web3Loader extends BaseLoader {
+  private _web3Contract: any;
+
+  public fromABI(abi: object, bytecode?: string, address?: string): any {
+    return new this.web3Contract(abi, address, { data: bytecode, from: this.defaultSender, gas: this.defaultGas });
+  }
+
+  protected get web3Contract(): any {
+    if (this._web3Contract === undefined) {
+      // If we only have a web3 provider, then we need to require web3-eth-contract
+      if (this.web3 === undefined) {
+        const lib = tryRequire('web3-eth-contract');
+        if (lib === undefined) {
+          throw new Error(
+            "Could not load package 'web3-eth-contract'. Please install it alongisde @openzeppelin/contract-loader.",
+          );
+        }
+        lib.setProvider(this.provider);
+        this._web3Contract = lib;
+      }
+      // Otherwise, we can use the web3.eth.Contract directly, and not require any extra deps
+      else {
+        this._web3Contract = this.web3.eth.Contract;
+      }
+    }
+
+    return this._web3Contract;
   }
+}
 
-  function fromABI(abi: object, bytecode = '') {
+export class TruffleLoader extends BaseLoader {
+  private _truffleContract: any;
+
+  public fromABI(abi: object, bytecode?: string, address?: string) {
     // eslint-disable-next-line @typescript-eslint/camelcase
-    const abstraction = truffleContract({ abi, unlinked_binary: bytecode });
-    abstraction.setProvider(provider);
-    abstraction.defaults({ from: defaultSender, gas: defaultGas });
+    const abstraction = this.truffleContract({ abi, unlinked_binary: bytecode });
+    abstraction.setProvider(this.provider);
+    abstraction.defaults({ from: this.defaultSender, gas: this.defaultGas });
 
+    if (address !== undefined) return new abstraction(address);
```

Comment:
Hm, I would've expected this to be `abstraction.at(address)` (which in turn is problematic because `at` returns a Promise).

I looked in [truffle's documentation](https://www.npmjs.com/package/@truffle/contract) and couldn't find anything about creating new contract instances using `new`, do you have a reference for that usage?

Tests are passing, so.. :man_shrugging:

Reply:
Yes, I did not use `at` specifically because of that - I didn't want to break this API by returning a promise instead of a contract.

As for documentation, you know what they say: code is the best documentation out there!